python/packages/autogen-core/src/autogen_core/memory/_base_memory.py

```diff
@@ -0,0 +1,95 @@
+from datetime import datetime
+from enum import Enum
+from typing import Any, Dict, List, Protocol, Union, runtime_checkable
+
+from pydantic import BaseModel, ConfigDict
+
+from .._cancellation_token import CancellationToken
+from .._image import Image
+from ..model_context import ChatCompletionContext
+
+
+class MemoryMimeType(Enum):
+    """Supported MIME types for memory content."""
+
+    TEXT = "text/plain"
+    JSON = "application/json"
+    MARKDOWN = "text/markdown"
+    IMAGE = "image/*"
+    BINARY = "application/octet-stream"
+
+
+ContentType = Union[str, bytes, Dict[str, Any], Image]
+
+
+class MemoryContent(BaseModel):
+    content: ContentType
+    mime_type: MemoryMimeType | str
+    metadata: Dict[str, Any] | None = None
+    timestamp: datetime | None = None
+    source: str | None = None
+    score: float | None = None
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+
+@runtime_checkable
+class Memory(Protocol):
+    """Protocol defining the interface for memory implementations."""
+
+    @property
+    def name(self) -> str | None:
+        """The name of this memory implementation."""
+        ...
+
+    async def update_context(
+        self,
+        model_context: ChatCompletionContext,
+    ) -> List[MemoryContent]:
+        """
+        Update the provided model context using relevant memory content.
+
+        Args:
+            model_context: The context to update.
+
+        Returns:
+            List of memory entries with relevance scores
+        """
+        ...
+
+    async def query(
+        self,
+        query: MemoryContent,
```

Comment:
Using `MemoryContent` for the query type seems odd. Can we start with `str | None`, in case of `str` automatically convert the query into a memory content under the hood for implementations. 

This gives us more flexibility in the future if we need to introduce a new type for query.

Reply:
Thanks.
The use of MemoryContent as query is based on design conversation .. see above from @colombod (which made sense!).

I also agree on supporting str (very common)

>I still think that content items should be with payload and mimetype, being a string doesn't explain what it is. is is a json text, plain text? content could also be structured objects I suppose 

<img width="758" alt="image" src="https://github.com/user-attachments/assets/f8e58862-6126-431f-beec-67211f85abd8" />

Reply:
Updated query to take str | MemoryContent